examples/transformers/README.md

```diff
@@ -0,0 +1,189 @@
+    get_cosine_schedule_with_warmup,
+    get_cosine_with_hard_restarts_schedule_with_warmup,
+    get_linear_schedule_with_warmup,
+    get_polynomial_decay_schedule_with_warmup,
+)
+
+from sparseml.pytorch.optim.manager import ScheduledModifierManager
+from sparseml.pytorch.optim.optimizer import ScheduledOptimizer
+from sparseml.pytorch.utils import ModuleExporter
+from sparseml.pytorch.helpers import any_str_or_regex_matches_param_name
+
+
+def get_sparsity_by_regex(module: Module, param_names: List[str]):
+    """
+    :param module: the module to get the matching layers and params from
+    :param param_names: a list of names or regex patterns to match with full parameter
+        paths. Regex patterns must be specified with the prefix 're:'
+    ::return sparsity: a float representing the percetage of zero-weight parameters
+    """
+    param_count = 0
+    param_set_to_zero = 0
+    for layer_name, layer in module.named_modules():
+        for param_name, param in layer.named_parameters():
+            if "." in param_name:  # skip parameters of nested layers
+                continue
+            full_param_name = "{}.{}".format(layer_name, param_name)
+            if any_str_or_regex_matches_param_name(full_param_name, param_names):
+                param_count += np.prod(param.size())
+                param_set_to_zero += torch.sum(param.data == 0)
+    sparsity = float(param_set_to_zero / param_count)
+    return sparsity
+
+
+def load_optimizer(model, args):
+    no_decay = ["bias", "LayerNorm.weight"]
+    optimizer_grouped_parameters = [
+        {
+            "params": [
+                p
+                for n, p in model.named_parameters()
+                if not any(nd in n for nd in no_decay)
+            ],
+            "weight_decay": args.weight_decay,
+        },
+        {
+            "params": [
+                p
+                for n, p in model.named_parameters()
+                if any(nd in n for nd in no_decay)
+            ],
+            "weight_decay": 0.0,
+        },
+    ]
+    optimizer_cls = AdamW
+    optimizer_kwargs = {
+        "betas": (args.adam_beta1, args.adam_beta2),
+        "eps": args.adam_epsilon,
+    }
+    optimizer_kwargs["lr"] = args.learning_rate
+    return optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
+```
+Use the code below to load sparseml optimizers
+```python
+## Neural Magic Integration here. 
+optim = load_optimizer(model, TrainingArguments) #We first create optimizers based on the method defined in transformers trainer class
+steps_per_epoch = math.ceil(len(datasets["train"]) / (training_args.n_gpu * training_args.per_device_train_batch_size))
+manager = ScheduledModifierManager.from_yaml(data_args.nm_prune_config) # Load a NM pruning config
+
+optim = ScheduledOptimizer(
+    optim, model, manager, steps_per_epoch=steps_per_epoch, loggers=None
+)
+```
+Modify the hugging face trainer to take the sparseml optimzier as shown below
```

Comment:
There's a new diff up from Ben that should make this integration a lot easier by modifying the optimizer in place: https://github.com/neuralmagic/sparseml/pull/77

Can we see if that will simplify this integration so we don't have to pass around our wrapped optimizer?

Reply:
rolled this back because we rolled back this fix.